tests/test_validation.py

```diff
@@ -0,0 +1,141 @@
+                             'template_empty_documentation.yml')
+    assert str(e.value) == "keyword(s) without any documentation: ['a_short_string']"
+
+    # keyword with invalid predicate
+    with pytest.raises(TemplateError) as e:
+        input_dict = _helper('template_errors',
+                             'input.yml',
+                             'template_invalid_predicate.yml')
+    assert str(e.value) == "error in predicate '0 < len(value) <= undefined' in keyword 'a_short_string'"
+
+
+def test_input_errors():
+
+    # unexpected keyword
+    with pytest.raises(InputError) as e:
+        input_dict = _helper('input_errors',
+                             'input_unexpected_keyword.yml',
+                             'template.yml')
+    assert str(e.value) == "found unexpected keyword(s): {'strange'}"
+
+    # unexpected section
+    with pytest.raises(InputError) as e:
+        input_dict = _helper('input_errors',
+                             'input_unexpected_section.yml',
+                             'template.yml')
+    assert str(e.value) == "found unexpected section(s): {'weird'}"
+
+    # keyword which has not default is not set
+    with pytest.raises(InputError) as e:
+        input_dict = _helper('input_errors',
+                             'input_missing_keyword.yml',
+                             'template.yml')
+    assert str(e.value) == "the following keyword(s) must be set: {'a_short_string'}"
+
+    # type errors
+    for file_name, error in [
+        ('input_type_error_bool.yml',
+         "incorrect type for keyword: 'some_feature', expected 'bool' type"),
+        ('input_type_error_float.yml',
+         "incorrect type for keyword: 'some_float', expected 'float' type"),
+        ('input_type_error_int.yml',
+         "incorrect type for keyword: 'some_number', expected 'int' type"),
+        ('input_type_error_list.yml',
+         "incorrect type for keyword: 'some_list', expected 'List[float]' type"),
+        ('input_type_error_str.yml',
+         "incorrect type for keyword: 'a_short_string', expected 'str' type"),
+                        ]:
+        with pytest.raises(InputError) as e:
+            input_dict = _helper('input_errors',
+                                 file_name,
+                                 'template.yml')
+        assert str(e.value) == error
+
+    # intra-keyword predicate validation
+    with pytest.raises(InputError) as e:
+        input_dict = _helper('input_errors',
+                             'input_predicate_intra.yml',
+                             'template.yml')
+    assert str(e.value) == 'predicate "value % 2 == 0" failed in keyword "another_number"'
+
+    # predicate validation across keywords
+    with pytest.raises(InputError) as e:
+        input_dict = _helper('input_errors',
+                             'input_predicate_cross.yml',
+                             'template.yml')
+    # in python < 3.6 the dict order is not guaranteed so we are not sure
```

Comment:
So in Py3.6+ `dict`-s are ordered by default? Good to know!

Reply:
Yes, order is preserved. This is very nice.